safe_transaction_service/events/services/queue_service.py

```diff
@@ -13,24 +13,44 @@
 logger = logging.getLogger(__name__)
 
 
-class QueueServiceProvider:
-    def __new__(cls):
-        if not hasattr(cls, "instance"):
-            if settings.EVENTS_QUEUE_URL:
-                if settings.EVENTS_QUEUE_ASYNC_CONNECTION:
-                    cls.instance = AsyncQueueService()
-                else:
-                    cls.instance = SyncQueueService()
-            else:
-                # Mock send_event to not configured host us is not mandatory configure a queue for events
-                cls.instance = MockedQueueService()
-                logger.warning("MockedQueueService is used")
-        return cls.instance
+def getQueueService():
+    if settings.EVENTS_QUEUE_URL:
+        return SyncQueueService()
+    else:
+        # Mock send_event to not configured host us is not mandatory configure a queue for events
+        return MockedQueueService()
+        logger.warning("MockedQueueService is used")
+
+
+class QueueServicePool:
+    """
+    Context manager to get a QueueService connection from the pool or create a new one and append it to the pool if all the
+    instances are taken. Very useful for gevent, as it is not safe to share one Pika connection across threads.
+    https://pika.readthedocs.io/en/stable/faq.html
+    # To avoid leave unsent messages in QueueService instances the queue service pool is rotating every QueueService in the list.
+    Use:
+    ```
+    with QueueServicePool() as queue_service:
+        queue_service...
+    ```
+    """
+
+    queue_service_pool = []
+
+    def __init__(self):
+        self.instance: QueueService
 
-    @classmethod
-    def del_singleton(cls):
-        if hasattr(cls, "instance"):
-            del cls.instance
+    def __enter__(self):
+        if self.queue_service_pool:
+            # If there are elements on the pool, take them
+            self.instance = self.queue_service_pool.pop()
+        else:
+            # If not, get a new client
+            self.instance = getQueueService()
+        return self.instance
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        self.queue_service_pool.insert(0, self.instance)
```

Comment:
@Uxio0 just mentions that this pool is going to increase the number of connections with general rabbitmq by pod. 
I could see during local tests that from indexer worker we can reach 29 connections simultaneously and because we are rotating every instance in the pool we are going to keep the connections opened. 
Do you think that this can be an issue? maybe we should limit the number of instances that the pool list can store and don´t store the new instances upper the limit if all the messages were sent.

Reply:
I don't like the structure of the service. Service should handle sending events and maintining a pool of RabbitMQ clients, having a pool of services look wrong to me. Also users of the service should not be aware that it's intended to use with a pool, that should be handled internally. Users should call `send_event` and all the magic should happen there

Reply:
I will handle to handle the pool of connections inside of the QueueService.

Reply:
Well, this [commit](https://github.com/safe-global/safe-transaction-service/pull/1818/commits/c2726e2a9bc105453a7e44f67996f05d0261fead) do a refactor separating the connection management from `QueueService` in a new class called `BrokerConnection`. 
At this way QueueService just handle the unsent events and the pool of `BrokerConnections`.